File: 이재진/미로탈출.py
import heapq
def djikstra(s,e,maps):
    dij = [[0,1], [1,0], [0,-1], [-1,0]]
    dis = [[float("inf")]*len(maps[0]) for _ in range(len(maps))]
    pq = []
    heapq.heappush(pq,[0,s[0],s[1]])
    dis[s[0]][s[1]] = 0

    while pq:
        w, ci, cj = heapq.heappop(pq)
        if ci==e[0] and cj==e[1]:
            return w
        if dis[ci][cj] >= w:
            for d in dij:
                ni, nj = ci+d[0], cj+d[1]
                new_w = w+1
                if 0<=ni<len(maps) and 0<=nj<len(maps[0]) and maps[ni][nj] != "X":
                    if dis[ni][nj] > new_w:
                        dis[ni][nj] = new_w
                        heapq.heappush(pq,[new_w,ni,nj])
    return False

def solution(maps):
    answer = 0
    start, end, lever = [],[],[]
    for i in range(len(maps)):
        for j in range(len(maps[i])):
            if maps[i][j] == "S":
                start = [i,j]
            elif maps[i][j] == "E":
                end = [i,j]
            elif maps[i][j] == "L":
                lever = [i,j]
    
    first = djikstra(start, lever, maps)
    if not first:
        return -1
    second = djikstra(lever, end, maps)
    if not second:
        return -1
    return first + second

File: 이재진/미로탈출.java
import java.util.*;

class Solution {
    
    public int[][] make_dis(String[] maps) {
        int[][] dis = new int[maps.length][maps[0].length()];
        for (int i=0;i<maps.length;i++){
            for (int j=0;j<maps[0].length();j++) {
                dis[i][j] = 9999999;
            }
        }
        return dis;
    }
    
    public int algo(String[] maps, int si, int sj, int ei, int ej) {
        int[][] dij = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        int[][] dis = make_dis(maps);
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        int[] first = {0,si,sj,0};
        pq.offer(first);
        while(!pq.isEmpty()) {
            int[] now = pq.poll();
            int w = now[0];
            int ci = now[1];
            int cj = now[2];
            int ans = now[3];
            if (ci == ei && cj == ej) {
                return ans;
            }
            if (w <= dis[ci][cj]) {
                for (int[] d:dij) {
                    int ni = ci+d[0];
                    int nj = cj+d[1];
                    if (0<=ni && ni<maps.length && 0<=nj && nj<maps[0].length() && maps[ni].charAt(nj) != 'X') {
                        if (dis[ni][nj] > w+1) {
                            dis[ni][nj] = w+1;
                            int[] next = {w+1, ni, nj, ans+1};
                            pq.offer(next);
                        }
                    }
                }
            }
        }
        return -1;
    }
    
    public int solution(String[] maps) {
        int answer = 0;
        int si = 0;
        int sj = 0;
        int ei = 0;
        int ej = 0;
        int li = 0;
        int lj = 0;
        
        for (int i=0;i<maps.length;i++) {
            for (int j=0;j<maps[i].length();j++) {
                if (maps[i].charAt(j) == 'S') {
                    si = i;
                    sj = j;
                } else if (maps[i].charAt(j) == 'E') {
                    ei = i;
                    ej = j;
                } else if (maps[i].charAt(j) == 'L') {
                    li = i;
                    lj = j;
                }
            }
        }
        if (algo(maps,si,sj,li,lj) == -1 || algo(maps,li,lj,ei,ej) == -1) {
            return -1;
        }
        
        answer += algo(maps,si,sj,li,lj);
        System.out.println(answer);
        answer += algo(maps,li,lj,ei,ej);
        return answer;
    }
}
